File: ups_server/logger.c
/* 
 * Modul printer definuje funkce pro výpis logů na konzoli.
 * 
 * Author: Petr Kozler
 */

#include "logger.h"
#include "config.h"
#include <error.h>
#include <errno.h>
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

/**
 * Vypíše aktuální datum a čas.
 * 
 * @param buf buffer pro uložení formátovaného řetězce s časovým údajem
 * @param success true v případě běžného záznamu, false v případě záznamu o chybě
 */
void print_datetime(char *buf, bool success) {
    time_t timer;
    struct tm* tm_info;

    time(&timer);
    tm_info = localtime(&timer);

    strftime(buf, MAX_STR_LENGHT, "%Y-%m-%d %H:%M:%S", tm_info);
    
    if (!success) {
        fprintf(stderr, "[%s] ", buf);
        
        return;
    }
    
    printf("[%s] ", buf);
}

/**
 * Vypíše na obrazovku zadanou hlášku a ukončí řádek.
 * 
 * @param format formát řetězce hlášky
 * @param ... argumenty řetězce hlášky
 */
void print_out(const char *format, ...) {
    char buf[MAX_STR_LENGHT];
    print_datetime(buf, true);
    
    va_list vargs;
    va_start(vargs, format);
    vsnprintf(buf, sizeof(char) * MAX_STR_LENGHT, format, vargs);
    va_end(vargs);
    
    printf("%s\n", buf);
}

/**
 * Vypíše záznam o zprávě přijaté od klienta.
 * 
 * @param msg řetězec zprávy
 * @param sock socket klienta
 * @param success true, pokud přenos proběhl v pořádku, jinak false
 */
void print_recv(const char *msg, int sock, bool success) {
    char buf[MAX_STR_LENGHT];
    
    if (!success) {
        print_datetime(buf, success);
        fprintf(stderr, "Server <-- Klient %d - Chyba příjmu: \"%s\"\n",
                sock, msg == NULL ? strerror(errno) : msg);
        
        return;
    }
    
    if (msg[0] == '\0') {
        return;
    }
    
    print_datetime(buf, success);
    printf("Server <-- Klient %d: \"%s\"\n", sock, msg);
}

/**
 * Vypíše záznam o zprávě zaslané klientovi.
 * 
 * @param msg řetězec zprávy
 * @param sock socket klienta
 * @param success true, pokud přenos proběhl v pořádku, jinak false
 */
void print_send(const char *msg, int sock, bool success) {
    char buf[MAX_STR_LENGHT];
    
    if (!success) {
        print_datetime(buf, success);
        fprintf(stderr, "Server --> Klient %d - Chyba odesílání: \"%s\"\n",
                sock, msg == NULL ? strerror(errno) : msg);
        
        return;
    }
    
    if (msg[0] == '\0') {
        return;
    }
    
    print_datetime(buf, success);
    printf("Server --> Klient %d: \"%s\"\n", sock, msg);
}

/**
 * Vypíše na obrazovku zadanou chybovou hlášku spolu s popisem chyby podle čísla
 * uloženého v konstantě errno a ukončí řádek.
 * 
 * @param format formát řetězce chybové hlášky
 * @param ... argumenty řetězce chybové hlášky
 */
void print_err(const char *format, ...) {
    char buf[MAX_STR_LENGHT];
    print_datetime(buf, false);
    
    va_list vargs;
    va_start(vargs, format);
    vsnprintf(buf, sizeof(char) * MAX_STR_LENGHT, format, vargs);
    va_end(vargs);
    
    fprintf(stderr, "%s: \"%s\"\n", buf, strerror(errno));
}

File: UpsClient/src/interaction/Logger.java
package interaction;

import java.text.SimpleDateFormat;
import java.util.Calendar;

/**
 * Třída Logger představuje objekt pro zápis logů, který je jako jedináček
 * přístupný pro použití z kterékoliv komponenty aplikace.
 * 
 * @author Petr Kozler
 */
public class Logger {
    
    /**
     * jediná instance loggeru
     */
    private static Logger instance;
    
    /**
     * Vytvoří objekt pro sestavení řetězce záznamu, který obsahuje
     * aktuální datum a čas.
     * 
     * @return objekt pro sestavení záznamu
     */
    private StringBuilder createLogBuilder() {
        return new StringBuilder("[").append(new SimpleDateFormat("yyyy-MM-dd HH:mm:ss").
                format(Calendar.getInstance().getTime())).append("] ");
    }
    
    /**
     * Vypíše na obrazovku zadanou hlášku a ukončí řádek.
     * 
     * @param format formát řetězce hlášky
     * @param args argumenty řetězce hlášky
     */
    public void printOut(String format, Object... args) {
        StringBuilder sb = createLogBuilder();
        sb.append(String.format(format, args));
        System.out.println(sb.toString());
    }
    
    /**
     * Vypíše záznam o zprávě přijaté od serveru.
     * 
     * @param msg řetězec zprávy
     */
    public void printRecv(String msg) {
        if (msg.isEmpty()) {
            return;
        }
        
        StringBuilder sb = createLogBuilder();
        sb.append("Klient <-- Server: \"").append(msg).append("\"");
        System.out.println(sb.toString());
    }
    
    /**
     * Vypíše záznam o zprávě zaslané serveru.
     * 
     * @param msg řetězec zprávy
     */
    public void printSend(String msg) {
        if (msg.isEmpty()) {
            return;
        }
        
        StringBuilder sb = createLogBuilder();
        sb.append("Klient --> Server: \"").append(msg).append("\"");
        System.out.println(sb.toString());
    }
    
    /**
     * Vypíše na obrazovku (do chybového výstupu) zadanou chybovou hlášku
     * a ukončí řádek.
     * 
     * @param format formát řetězce hlášky
     * @param args argumenty řetězce hlášky
     */
    public void printErr(String format, Object... args) {
        StringBuilder sb = createLogBuilder();
        sb.append(String.format(format, args));
        System.err.println(sb.toString());
    }
    
    /**
     * Vrátí (a při prvním volání nejprve vytvoří) jedinou instanci loggeru
     * přístupnou pro kteroukoliv komponentu aplikace.
     * 
     * @return instance loggeru
     */
    public static Logger getInstance() {
        if (instance == null) {
            instance = new Logger();
        }
        
        return instance;
    }
    
    /**
     * Privátní konstruktor pro zamezení vytvoření dalších instancí.
     */
    private Logger() {
        // žádný kód
    }
    
}
